### ISSUE TITLE
Launching Hardware Configuration Wizard Unloads All Devices

### ISSUE THREAD

**Issue by gselzer** [gselzer/pymmcore-plus-sandbox#29]
This seems to be reproducible. Likely an issue upstream in pymmcore-widgets, but also might be easily avoided with some changes here.

Steps to reproduce:
1. Load the Demo config
2. Snap an image
3. **(Optional)** `Device -> Device Property Browser` to confirm that all demo properties exist.
4. `Device -> Hardware Configuration Wizard` to launch the wizard. Cancel it right away. Note that `DEBUG:pymmcore-plus:Unloading all devices` prints to console.
5. `Device -> Device Property Browser` to confirm that only the core properties remain.

Ideally, cancelling the wizard would retain the configuration before the wizard was started (right?)

**Comment by marktsuchida** [comment]
I think in MMStudio the previous config file is reloaded if the wizard is canceled at any point. It would certainly be ideal if the devices are never unloaded in the first place when canceling before actually doing anything.

**Comment by gselzer** [comment]
Update: I tried to get around this by saving the existing state in a pymmcore-plus `Microscope` object, restoring it if the wizard is rejected:

```python
        existing = Microscope.create_from_core(self._mmc)
        wizard = ConfigWizard(parent=None, core=self._mmc)
        wizard.exec()
        if wizard.rejected:
            existing.apply_to_core(self._mmc)
```

This unfortunately hits a *nasty* bug within the `PresetsWidget` of pymmcore-widgets, though, where config group properties are dropped. cc @tlambert03

**Comment by tlambert03** [comment]
yeah, the behavior of that widget was largely modeled on the behavior of MMStudio, I'm pretty sure it also unloads all devices if you proceed into the hardware config wizard... the only difference here is that closing that widget doesn't restore whatever it was before the widget was loaded (since the widget wasn't created in the context of an app).

note that rather than the workaround you used above, you can use `self._mmc.loadSystemConfiguration(self._mmc.systemConfigurationFile())` 
see: https://pymmcore-plus.github.io/pymmcore-plus/api/cmmcoreplus/#pymmcore_plus.core._mmcore_plus.CMMCorePlus.systemConfigurationFile

`systemConfigurationFile()` returns the path to the last-loaded config file.  That's something that MMStudio hangs on to, but we add a core method in pymmcore-plus

**Comment by gselzer** [comment]
> `systemConfigurationFile()` returns the path to the last-loaded config file. That's something that MMStudio hangs on to, but we add a core method in pymmcore-plus

Ah, good to know, thanks @tlambert03!

What are your opinions on adding something like this upstream into the widget's code?

**Comment by gselzer** [comment]
FYI I've fixed the issue with ac044f0, however I'm going to leave this issue open until we decide what to do about any upstream fix.

**Comment by tlambert03** [comment]
> What are your opinions on adding something like this upstream into the widget's code?

that seems reasonable to me.  since the widget definitely modifies that state of core, I see no problem with having it revert to the state it found it if the user cancels

**Comment by gselzer** [comment]
> > What are your opinions on adding something like this upstream into the widget's code?
> 
> that seems reasonable to me. since the widget definitely modifies that state of core, I see no problem with having it revert to the state it found it if the user cancels

Cool, I'll make a PR!

**Comment by gselzer** [comment]
See pymmcore-plus/pymmcore-widgets#359
